### DISCUSSION TITLE
State of TypeScript and Fastify

### DISCUSSION THREAD

**Discussion by Ethan-Arrowood** [fastify/fastify#2877] [Ideas]
Hi folks! I wanted to open a discussion about the state of TypeScript in fastify. We've had many new contributes join us helping improve TypeScript support across the ecosystem. Here are some questions to get us started.

1. What are some pain points in the current definitions you'd like to be fixed?
2. What aspect of the definitions do you enjoy and not enjoy using?
3. What external tools or utilities have you used to help you use fastify with TypeScript?
4. What server implementation patterns have you found successful when using fastify with TypeScript?

Furthermore, if there is something else you'd like to see discussed feel free to ask the question! Everyone (users & contributors) are encourage to reply (yes, even the folks who don't use TypeScript but want better support for things like code-hints in their editor) ❤️

**Comment by L2jLiga** [comment]
Hi,

> 1. What are some pain points in the current definitions you'd like to be fixed?
> 2. What aspect of the definitions do you enjoy and not enjoy using?

The biggest pain point for me is generics and hooks, such things like support in hooks arrays of functions, or just functions which may use `done` or `Promise`

<details>
<summary>Example</summary>

```typescript
{
  preValidation: ((
      request: FastifyRequest<{ Querystring: { fileName: string | string[] } }>, _reply: FastifyReply, done: () => void,
    ) => {
      if (!Array.isArray(request.query.fileName)) request.query.fileName = [request.query.fileName];
      done();
    }) as preValidationHookHandler
}
```

without casting it as `preValidationHookHandler explicitely I'm facing with error

```
TS2322: Type '(request: FastifyRequest<{ Querystring: { fileName: string | string[]; };}>, _reply: FastifyReply, done: () => void) => void' is not assignable to type 'preValidationHookHandler<Server, IncomingMessage, ServerResponse, RouteGenericInterface, unknown> | preValidationHookHandler<...>[] | undefined'.
   Type '(request: FastifyRequest<{ Querystring: { fileName: string | string[]; };}>, _reply: FastifyReply, done: () => void) => void' is not assignable to type 'preValidationHookHandler<Server, IncomingMessage, ServerResponse, RouteGenericInterface, unknown>'.
     Types of parameters 'request' and 'request' are incompatible.       Type 'FastifyRequest<RouteGenericInterface, Server, IncomingMessage>' is not assignable to type 'FastifyRequest<{ Querystring: { fileName: string | string[]; }; }, Server, IncomingMessage>'.
         Type 'RouteGenericInterface' is not assignable to type '{ Querystring: { fileName: string | string[]; }; }'.
           Types of property 'Querystring' are incompatible.
             Type 'unknown' is not assignable to type '{ fileName: string | string[]; }'.

  route.d.ts(39, 3): The expected type comes from property 'preValidation' which is declared here on type 'RouteShorthandOptions<Server, IncomingMessage, ServerResponse, RouteGenericInterface, unknown>'
```
</details>


> 3. What external tools or utilities have you used to help you use fastify with TypeScript?
> 4. What server implementation patterns have you found successful when using fastify with TypeScript?

I've written own library to deal with it, so my solution is to use [`fastify-decorators`](https://github.com/L2jLiga/fastify-decorators) combined it with [`typebox`](https://github.com/sinclairzx81/typebox) and [`typeorm`](http://typeorm.io/), I even have own [plugin](https://github.com/L2jLiga/fastify-decorators-plugin) for my library 😄 

In general my TS code looks like several controllers, services, entities, I also use separated files for schemas and their TS types :)

<details>

<summary>Example</summary>

*schemas.ts*:
```typescript
import { Type } from '@sinclair/typebox'

export const filesSchema = Type.Object({
  fileName: Type.Array(Type.String()),
}, {
  examples: [
    {
      name: 'Single file',
      summary: 'Single file',
      value: { fileName: 'sample.xd' },
    },
    {
      name: 'Multiple files',
      summary: 'Multiple files',
      value: { fileName: ['sample.xd', 'sample2.xd'] },
    },
  ],
})

export type Files = Static<typeof filesSchema>
```

*FileController.ts*:
```typescript
import { Controller, GET } from 'fastify-decorators'
import { filesSchema, Files } from './schemas.js'

@Controller('/files')
export class FilesController {
  @GET('/', { schema: { query: filesSchema } })
  async getFiles(request: FastifyRequest<{ Querystring: Files  }>) {
    // async stuff here
  }
}
```

</details>

**Reply by Ethan-Arrowood** [reply]
I agree that is not a good experience with the hooks. I think the situation with generics in the type system needs an overall improvement.

**Reply by Ethan-Arrowood** [reply]
The decorator stuff is great. Honestly I'm starting to think that that model is best for a TypeScript x Fastify app. Nest.js has tremendous success with the pattern.

**Comment by airhorns** [comment]
A big pain point for me is that you need to use module augmentation to add types for the decorators that you add. I totally understand why it is necessary, but it is arcane syntax that I would guess many aren't familiar with, and requires a pretty deep understanding of the TS module concepts to get right. 

Another one is that the types aren't aware of the encapsulation context, so you can't quite trust the types of the decorators anyways. If you decorate within an encapsulation context, you either have to type all `FastifyRequest`s with that type which is a lie outside the context, or not decorate at all, which is a lie inside the context. It is also no fun to decorate twice with the same plugin but using a different type, like say `fastify-passport` with two different user logins and user types, one for your actual users and one for your staff accessing an admin area. 

Having a blessed path for maintaining JSON schemas and derived types without writing them out twice would be dandy as well.

**Reply by Ethan-Arrowood** [reply]
Yesh module augmentation is definitely a bandaid solution for a much harder problem. I'd love to see another way to get encapsulation to work without it. TypeScript decorator pattern might be the best solution honestly.

**Reply by zekth** [reply]
I agree with this

**Reply by RafaelGSS** [reply]
Strongly agree. The module augmentation based on encapsulation context is still a pending feature.

**Reply by mcollina** [reply]
Maybe we could add some sort of different plugin syntax where this is supported. I don't have many ideas but I would be very keen to see this solved (however it's outside of my TS skills).

Maybe we could add a `server.ensure(plugin)` that will make sure all the decorators are loaded properly.

**Reply by spiffytech** [reply]
This problem would be improved if there was a way to pass around a `FastifyInstance` type that used a custom request type instead of the hardcoded `FastifyRequest`. I'm already passing around my server instance to all the subsystems that need to set up routes and decorators. Having those subsystems expect something like `FastifyInstance<{myCustomProp: Foo} & FastifyRequest> & {myServerProp: Bar}` would be a big improvement, but the current types don't seem to allow this.

**Reply by Eomm** [reply]
Did you tried the next branch too?
The types has been improved a lot there and it will be release very soon within fastify v4

**Reply by spiffytech** [reply]
I tried pulling in the `next` branch, but all the Fastify plugins I use demand fastify@3.x so I didn't get very far with that.

I checked the types on `next` and it looks like encapsulated/scoped extensions to `FastifyRequest` are not presently supported. Both `FastifyInstance` and the route handler types hardcode `FastifyRequest` with no generic parameter to override that, so I don't see a way pass around a custom request type that my routes pick up.

**Reply by RafaelGSS** [reply]
It's a pending feature. See: https://github.com/fastify/fastify/issues/3529

**Reply by rubenferreira97** [reply]
As far as I know is not possible to scope TS declaration merging.

Please acknowledge these words as constructive criticism, but didn't fastify team acknowledge in February (via feedback) that declaration merging is a "band aid solution" and shouldn't already have looked for another architectural/structural solution? Declaration merging IS the problem of unscoped instances. I am aware that fastify wasn't build on TypeScript but I think it can do better.

Another issue for developers is they can't specify/build a custom plugin with custom types (Eg: Request, Reply, etc...). We have so many generics but then for some "basic" stuff we have no control.

Fastify is awesome, but needless to say there is always room for improvement.

**Reply by RafaelGSS** [reply]
> Fastify is awesome, but needless to say there is always room for improvement.

Sure thing. Feel free to open an issue or use the above one linked to discuss a better approach for that. It would be certainly useful.

**Reply by mcollina** [reply]
> Please acknowledge these words as constructive criticism, but didn't fastify team acknowledge in February (via feedback) that declaration merging is a "band aid solution" and shouldn't already have looked for another architectural/structural solution?

No, would you like to look for another solution? As always, all contributions are welcome here. I would love to read your thoughts and your designs.

You can express criticism on the technical front and we often have some heated debates between ourselves too - it's part of OSS. However criticizing how volunteers decide to spend their own time is not tolerated. Those comments will quickly antagonize this community. Be constructive and do not demand free work.

**Comment by delvedor** [comment]
> What are some pain points in the current definitions you'd like to be fixed?

Having an out-of-the-box system to translate json schemas to types would be awesome.
The bets solution would be to offer one that just works™ and a way to inject your own, in the same way we are doing for validations and ajv.

**Reply by Ethan-Arrowood** [reply]
I love this idea!

**Comment by Ethan-Arrowood** [comment]
Check out this encapsulation solution I came up with: 

```ts
type Server = {
  decorate: (key: string, value: any) => void
  get: (url: string, handler: (request: unknown, response: unknown) => void) => void
}
type Options = unknown

type FPlugin<S = unknown> = (server: Server & S, options: Options) => Promise<void>
type EncapsulatedFPlugin<P extends FPlugin<any>> = FPlugin<Parameters<P>[0]>

type AuthDecoration = {
  foo: string
}

const authPlugin: FPlugin<AuthDecoration> = async (server, options) => {
  server.decorate('foo', 'bar')
}

const privateRoute: EncapsulatedFPlugin<typeof authPlugin> = async (server, options) => {
  server.get('/', (request, response) => {
    server.foo
  })
}
```

[TypeScript Playground](https://www.typescriptlang.org/play?ts=4.2.0-beta&ssl=1&ssc=1&pln=22&pc=2#code/C4TwDgpgBAyhBOA3BUC8UDeAoKUAmEAxgPbwCGwEAXFABQDWEINAzsPAJYB2A5gDRREZADYBXalDJcQASjQA+QcQ54cUHhGA1ao+MNbtu-KAAspeYQm3wIAR3FsaorvS7EA7lwE2WYYlxYJZ1cPLjlURURlPHDI6KwAXyxQSCgAeTBgDn8WNChgt08sZPBoADEABTEebgAeGDyC0MV0WkCkK1gEZHgoADJYAWJM7ICaDKyc2KgK+GIAWw5A2qiVeRLUgFEuQjIwFlFhCgg8Suq6iqgIAA9KLjxcs9EarlqpEHkWqCeX2oqycjzTQIFh-eQAbQADABddYbaAAQVEwBMABEiKQKKM8thcAAzYjEAycXiJYokALASTIkxVZ7cGg-OpIlHokjkSZcL5kFggHZ0do9IYjKYKTBqQUIAB0BHZx1oAHICcQFQIFQAjAEKmRkrAUthQMCcISUABKxGREm2u32h2OpzpvxSEGIeOpKMd3G5vP5bW6CGFnJY01xUEl8ClGmAioA9Kq6DZ7BA2N5k34AhAQ2pcOGpcq1AkdQkgA)

This could be potentially useful for properly encapsulated fastify plugin types without the need of TypeScript decorator syntax

**Reply by Ethan-Arrowood** [reply]
My idea with this is that whenever I build fastify apps, I already put all of my plugins and routes into their files and functions so that the encapsulation organization makes more sense implicitly. I think with a TypeScript utility like this, it would be a better experience for TS folks than the existing augmentation method.

**Comment by lachlancollins** [comment]
Hi Fastify team, love the work you're doing!

I just have a suggestion based on my experience using  documentation for other projects - I'm wondering if it would be possible to integrate TS info into the each of the docs rather than having it as its own separate page (e.g. including info on how to type plugins within the plugins documentation)?

Also as a bonus suggestion - I've never worked with metalsmith so I'm not sure how hard it would be to add this functionality, but having some sort of code toggle between JS and TS examples would be amazing :) The example below is from the [Redux Toolkit docs](https://redux-toolkit.js.org/tutorials/rtk-query).

![image](https://user-images.githubusercontent.com/1667261/142309794-88c996ce-40f9-42e9-ada7-465b2b9f00ac.png)

**Reply by Eomm** [reply]
Would you like to join this discussion: https://github.com/fastify/website/issues/291 ?
